src/codegeneration/SuperTransformer.js

```diff
@@ -12,159 +12,298 @@
+    if (this.home_ === null) {
+      this.home_ =
+          createIdentifierExpression(createIdentifierToken(this.tempName));
+    }
+    return this.home_;
   }
-  transformFunctionBody(tree) {
-    return tree;
+}
+
+/**
+ * Used to keep track of the name of the variable representing the class object
+ * as we transform the tree recursively.
+ */
+class ClassState extends State {
+  constructor(transformer, tree) {
+    let home = null;
+    if (tree.name !== null) {
+      home = createIdentifierExpression(tree.name.identifierToken);
+    }
+    super(transformer, home);
+    this.name_ = tree.name;
+  }
+
+  get name() {
+    if (this.name_ !== null) return this.name_;
+    if (this.hasSuper) {
+      return createBindingIdentifier(this.home.identifierToken);
+    }
+    return null;
   }
 }
 
 /**
- * Transforms super expressions in function bodies.
+ * Used to keep track of the name of the variable representing the prototype
+ * object as we transform the tree recursively.
+ */
+class PrototypeState extends State {
+  constructor(transformer, classState) {
+    super(transformer, null);
+    this.classState = classState;
+  }
+
+  get home() {
+    let ident = this.classState.home;
+    return new MemberExpression(null, ident,
+                                createIdentifierToken('prototype'));
+  }
+}
+
+/**
+ * Transforms super in object literals and class literals.
+ *
+ * For object literals we do something like this:
+ *
+ *   {
+ *     m() { super.x(); }
+ *   }
+ *
+ * =>
+ *
+ *   $tmp = { m() { $traceurRuntime.superGet(this, $tmp, "x").call(this); } }
+ *
+ * For classes we just ensure that the class has a name
+ *
+ *   class {
+ *     m() { super.x() }
+ *   }
  *
- *   super.x  =>  superGet(this, proto, 'x')
- *   super.x = expr  =>  superSet(this, proto, 'x', expr)
- *   super.x(1, 2)  =>  superGet(this, proto, 'x').call(this, 1, 2)
+ * =>
  *
- * This also transforms super.x++, ++super.x and super.x += expr
- * into forms that use the runtime functions.
+ *   class $tmp {
+ *     m() {
+ *       $traceurRuntime.superGet(this, $tmp.prototype, "x").call(this);
+ *     }
+ *   }
  */
-export class SuperTransformer extends ParseTreeTransformer {
-  /**
-   * @param {TempVarTransformer} tempVarTransformer
-   * @param {ParseTree} protoName
-   * @param {string} thisName The name of the saved 'this' var
-   * @param {ParseTree} internalName The name of the save class binding.
-   */
-  constructor(tempVarTransformer, protoName, thisName, internalName) {
-    super();
-    this.tempVarTransformer_ = tempVarTransformer;
-    this.protoName_ = protoName;
-    this.internalName_ = internalName;
-    this.superCount_ = 0;
-    this.thisVar_ = createIdentifierExpression(thisName);
-    this.inNestedFunc_ = 0;
-    this.nestedSuperCount_ = 0;
+export class SuperTransformer extends TempVarTransformer {
+  constructor(identifierGenerator, reporter, options) {
+    super(identifierGenerator, reporter, options);
+    this.stateStack_ = [];
```

Comment:
This datastructure is confusing me. I guess it is really a stack of pairs of (classState, prototypeState) so that `peekState()` is `peekPrototypeState()` and `this.stateStack_[this.stateStack_.length - 2];` is `peekClassState()`. But then the `popState()` for static accessors doing something magical. Perhaps you can comment on what it intended.